freud/util.pyx

```diff
@@ -0,0 +1,158 @@
+# Copyright (c) 2010-2019 The Regents of the University of Michigan
+# This file is from the freud project, released under the BSD 3-Clause License.
+
+import sys
+from functools import wraps
+
+from cython.operator cimport dereference
+cimport freud.util
+
+cimport numpy as np
+
+# numpy must be initialized. When using numpy from C or Cython you must
+# _always_ do that, or you will have segfaults
+np.import_array()
+
+
+cdef class ManagedArrayManager:
+    """Class responsible for synchronizing ownership between two ManagedArray
+    instances.
+
+    The purpose of this container is to minimize memory copies while avoiding
+    reference rot. Users obtaining numpy arrays as outputs from a compute class
+    have a reasonable expectation that these arrays will remain valid after the
+    compute class recomputes or goes out of scope. To enforce this requirement,
+    all such data members should be stored in using the ManagedArray class in
+    C++ and then synchronized using a ManagedArrayManager in Python. This class
+    retains a Python-level version of the array that maintains ownership of the
+    data between compute calls and only relinquishes ownership to the C++ class
+    if no outstanding references to the data exist in Python.
+
+    This class should always be initialized using the static factory
+    :meth:`~ManagedArrayManager.init` method, which creates the Python copy of
+    a ManagedArray provided the instance member of the underlying C++ compute
+    class.
+
+    .. moduleauthor:: Vyas Ramasubramani <vramasub@umich.edu>
+    """
+
+    def __cinit__(self, arr_type, typenum):
+        # This class should generally be initialized via the factory "init"
+        # function, but some logic is included here for ease of use.
+        self.data_type = arr_type
+        self.var_typenum = typenum
+        self._shape = tuple()
+        self.thisptr.null_ptr = NULL
+
+    @property
+    def shape(self):
+        return self._shape
+
+    @shape.setter
+    def shape(self, shape):
+        """Set the shape of the output numpy array when :meth:`numpy.asarray`
+        is called."""
+        self._shape = shape
+
+    def __dealloc__(self):
+        if self.var_typenum == np.NPY_UINT32:
+            del self.thisptr.uint_ptr
+
+    def __array__(self):
+        """Convert the underlying data array into a read-only numpy array."""
+        if self.shape == tuple():
+            raise ValueError("You must specify the shape of the numpy array "
+                             "to be created by calling set_shape.")
+        cdef unsigned int ndim = len(self.shape)
+
+        # These arrays must be allocated at compile time, so we make separate
+        # branches for different dimensionalities.
```

Comment:
Is there a better way? This is gross. 😕

Reply:
Lol not that I'm aware of. What would you suggest?

Reply:
@vyasr I thought some more about this -- try creating it as a one-dimensional array with size `np.prod(self.shape)` and calling `arr.reshape(self.shape)` after you've created the array object? This returns a view, so it should avoid copies.

Reply:
Good call, this definitely cleans up the code in `__array__`.